variable/variable.cpp

```diff
@@ -15,198 +15,133 @@
+  for (scipp::index i = 0; i < dims().ndim(); ++i)
+    ordered.emplace_back(dims().label(i), strides()[i]);
+  std::reverse(ordered.begin(), ordered.end());
+  std::stable_sort(
+      ordered.begin(), ordered.end(),
+      [](const auto &a, const auto &b) { return a.second < b.second; });
+  if (!ordered.empty() && ordered.front().second > 1) {
+    dataDims.add(Dim::X, ordered.front().second);
+    dataDims.relabel(0, Dim::Invalid);
+  }
+  for (scipp::index i = 0; i < scipp::size(ordered); ++i) {
+    const auto &[dim, stride] = ordered[i];
+    dataDims.add(dim, std::max(dims()[dim], ((i == scipp::size(ordered) - 1)
+                                                 ? m_object->size()
+                                                 : ordered[i + 1].second)) /
+                          std::max(scipp::index(1), stride));
+  }
+  return {m_offset, dims(), dataDims, {}};
 }
 
-VariableConstView
-VariableConstView::transpose(const std::vector<Dim> &order) const {
-  auto transposed(*this);
-  transposed.m_dims = core::transpose(dims(), order);
-  return transposed;
+Variable Variable::slice(const Slice slice) const {
+  Variable out(*this);
+  const auto dim = slice.dim();
+  const auto begin = slice.begin();
+  const auto end = slice.end();
+  const auto index = out.m_dims.index(dim);
+  out.m_offset += begin * m_strides[index];
+  if (end == -1) {
+    out.m_strides.erase(index);
+    out.m_dims.erase(dim);
+  } else
+    out.m_dims.resize(dim, end - begin);
+  return out;
 }
 
-VariableView VariableView::transpose(const std::vector<Dim> &order) const {
+Variable Variable::transpose(const std::vector<Dim> &order) const {
   auto transposed(*this);
+  Dimensions tmp = dims();
+  for (scipp::index i = 0; i < tmp.ndim(); ++i)
+    tmp.resize(i, strides()[i]);
+  tmp = core::transpose(tmp, order);
+  transposed.m_strides = Strides(tmp.shape());
   transposed.m_dims = core::transpose(dims(), order);
   return transposed;
 }
 
-std::vector<scipp::index> VariableConstView::strides() const {
-  const auto parent = m_variable->dims();
-  std::vector<scipp::index> strides;
-  for (const auto &label : parent.labels())
-    if (dims().contains(label))
-      strides.emplace_back(parent.offset(label));
-  return strides;
-}
-
-bool VariableConstView::is_trivial() const noexcept {
-  return m_variable && m_offset == 0 && m_dims == m_variable->dims() &&
-         m_dataDims == m_variable->dims();
-}
-
 void Variable::rename(const Dim from, const Dim to) {
   if (dims().contains(from))
-    data().m_dimensions.relabel(dims().index(from), to);
+    m_dims.relabel(dims().index(from), to);
 }
 
-/// Rename dims of a view. Does NOT rename dims of the underlying variable.
-void VariableConstView::rename(const Dim from, const Dim to) {
-  if (dims().contains(from)) {
-    m_dims.relabel(m_dims.index(from), to);
-    m_dataDims.relabel(m_dataDims.index(from), to);
-  }
+bool Variable::is_slice() const {
+  // TODO Is this condition sufficient?
```

Comment:
Remove "todo"? Looks safe to me...

Reply:
I am not convinced. What if we slice out a dimension but then broadcast into another of same length? I have the feeling we must also somehow include the strides here.

Reply:
> Do you want to detect whether the Variable views the entire buffer

Yes. I need to check whether the variable views *every single element* of the buffer. Order is *not* relevant.

What about the broadcasting case though? So you have (x:4, y:3) -> slice (x:4) -> broadcast (z:3, x:4).

Reply:
Can this share the buffer?
And like I said, even if you detect this case in the new variable, the old one can still change the unit. So the question would become, which object ha the original view of the data; Not which one has a complete view. I'm unsure if this is a useful criterion.

Reply:
Yes, I think this could share the buffer. All you need is something like:
```
offset = 0
dims = (z:4, y:3)
strides = (z:0, y:1)
```
I think?

There is no original view. This is imply about the definition/invariant "all elements in a variable have the same unit".

Reply:
I thought broadcasting on the Variable level was implemented by actually repeating the data in the new dim. Is the method you describe new or has it been around before?

With your example, I agree, we should flag this as a slice. Is it enough to just check for `stride[i]==0` or are there any other situations that are not covered by this and the size and offset check?

Reply:
> And like I said, even if you detect this case in the new variable, the old one can still change the unit. So the question would become, which object ha the original view of the data; Not which one has a complete view. I'm unsure if this is a useful criterion.

To give an example of what this protects us from:
```python
for i in range(4):
   var['x', i] *= 1.2 * sc.Unit('m')
```
would multiply `var.unit` by `m**4`, so modifying the unit via a partial view must be prevented.

Reply:
Is this not covered by the existing size check?

Reply:
> I thought broadcasting on the Variable level was implemented by actually repeating the data in the new dim. Is the method you describe new or has it been around before?

This may be what the current implementation is doing, yes, but with the more flexible stride handling we may want to change `broadcast` to just set corresponding strides (maybe optionally).

Reply:
> Is this not covered by the existing size check?

Yes it is, I was answering to your question about whether it makes sense at all, since we don't know what the "original" view is.

Reply:
It sounds like we misunderstood each other. My point was not about the existing check, I agree that this protection is useful. I was assuming that broadcasting works by actually including the whole underlying buffer and not by setting a stride to 0. In that case, the result of a broadcast (with suitable parameters) would include the entire buffer, not just part of it as in the case of slicing. So that would have shifted the logic away from not changing a unit because we only have part of the buffer to not changing the unit because the variable is derived from another.

Reply:
Just tried this with `numpy`, it also uses stride=0 for this:
```python
import numpy as np
a = np.random.rand(4,3)
b = np.broadcast_to(a[1], (4,3))
b.strides
```
gives `(0, 8)` (note strides are in Byte, so 8 moves to the next float64).